File: sample/src/main/java/contacts/sample/cheatsheet/basics/java/QueryContactsAdvancedActivity.java
package contacts.sample.cheatsheet.basics.java;

import static contacts.core.WhereKt.*;
import static contacts.core.util.ContactLookupKeyKt.lookupKeyIn;

import android.accounts.Account;
import android.app.Activity;

import java.util.List;

import contacts.core.ContactsFactory;
import contacts.core.Fields;
import contacts.core.entities.Contact;

public class QueryContactsAdvancedActivity extends Activity {
    @Nullable
    public Contact getContactById(Long contactId) {
        Query.Result result = contactsFactory
                .query()
                .where(equalTo(Fields.Contact.Id, contactId))
                .limit(1)
                .find();
        return !result.isEmpty() ? result.get(0) : null;
    }

    List<Contact> getContactByLookupKey(String lookupKey) {
        return ContactsFactory.create(this)
                .query()
                .where(
                        lookupKeyIn(Fields.Contact, lookupKey)
                )
                .find();
    }

    List<Contact> getAllContactsForAGoogleAccount() {
        return ContactsFactory.create(this)
                .query()
                .accounts(new Account("email@gmail.com", "com.google"))
                .find();
    }

    List<Contact> getOnlyFavoriteContacts() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        equalTo(Fields.Contact.Options.Starred, true)
                )
                .find();
    }

    List<Contact> getContactsPartiallyMatchingDisplayName() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        contains(Fields.Contact.DisplayNamePrimary, "alex")
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOneGmailEmail() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        endsWith(Fields.Email.Address, "@gmail.com")
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOnePhoneNumber() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        equalTo(Fields.Contact.HasPhoneNumber, true)
                        // isNotNullOrEmpty(Fields.Phone.Number) this works too but the above is more optimized
                )
                .find();
    }

    List<Contact> getContactsWithAtLeastOnePhoneNumberAndEmail() {
        return ContactsFactory.create(this)
                .query()
                .where(
                        and(
                                isNotNullOrEmpty(Fields.Phone.Number),
                                // or equalTo(Fields.Contact.HasPhoneNumber, true),
                                isNotNullOrEmpty(Fields.Email.Address)
                        )
                )
                .find();
    }
}

File: sample/src/main/java/contacts/sample/cheatsheet/basics/kotlin/QueryContactsAdvancedActivity.kt
package contacts.sample.cheatsheet.basics.kotlin

import android.accounts.Account
import android.app.Activity
import contacts.core.*
import contacts.core.entities.Contact
import contacts.core.util.lookupKeyIn

class QueryContactsAdvancedActivity : Activity() {

    fun getContactById(contactId: Long): Contact? = Contacts(this)
        .query()
        .where { Contact.Id equalTo contactId }
        .find()
        .firstOrNull()

    fun getContactByLookupKey(lookupKey: String, contactId: Long): List<Contact> = Contacts(this)
        .query()
        .where { Contact.lookupKeyIn(lookupKey)?.or(Contact.Id equalTo contactId) }
        .find()

    fun getAllContactsForAGoogleAccount(): List<Contact> = Contacts(this)
        .query()
        .accounts(Account("email@gmail.com", "com.google"))
        .find()

    fun getOnlyFavoriteContacts(): List<Contact> = Contacts(this)
        .query()
        .where {
            Contact.Options.Starred equalTo true
        }
        .find()

    fun getContactsPartiallyMatchingDisplayName(): List<Contact> = Contacts(this)
        .query()
        .where {
            Contact.DisplayNamePrimary contains "alex"
        }
        .find()

    fun getContactsWithAtLeastOneGmailEmail(): List<Contact> = Contacts(this)
        .query()
        .where {
            Email.Address endsWith "@gmail.com"
        }
        .find()

    fun getContactsWithAtLeastOnePhoneNumber(): List<Contact> = Contacts(this)
        .query()
        .where {
            Contact.HasPhoneNumber equalTo true
            // Phone.Number.isNotNullOrEmpty() this works too but the above is more optimized
        }
        .find()

    fun getContactsWithAtLeastOnePhoneNumberAndEmail(): List<Contact> = Contacts(this)
        .query()
        .where {
            Phone.Number.isNotNullOrEmpty() and Email.Address.isNotNullOrEmpty()
            // or Contact.HasPhoneNumber equalTo true and Email.Address.isNotNullOrEmpty()
        }
        .find()
}